### PR TITLE
feat(new_metrics): implement the volatile counter

### PR DESCRIPTION
**Pull request by empiredan**
https://github.com/apache/incubator-pegasus/issues/933

### CODE REVIEW

File: include/dsn/utility/metrics.h [hunk 1]
```diff
@@ -359,4 +369,35 @@ using concurrent_counter_ptr = counter_ptr<concurrent_long_adder>;
 template <typename Adder = striped_long_adder>
 using counter_prototype = metric_prototype_with<counter<Adder>>;
 
+// A volatile counter is nearly the same as a counter, except that fetch_and_reset() is provided
+// to reset the counter atomically after its value is fetched.
+// The reason why the volatile counter is introduced is that sometimes "recent" counters are
```
Comment (acelyc111) [inline include/dsn/utility/metrics.h:374]: Sounds it also similar to gauge type, how to distinguish gauge, counter, and volatile counter?

Comment (empiredan) [reply include/dsn/utility/metrics.h:374]: > Sounds it also similar to gauge type, how to distinguish gauge, counter, and volatile counter?

- The *gauge* is an instantaneous measurement that just provides `set()` and `get()`. The value fetched from it is what it's recently set. It's can be set at any value that arbitrarily goes up and down.
- The *counter*, as its name implies, starts out at 0 and provides `increment()` and `decrement()` to accumulate.
- The *volatile counter*, based on the general *counter*, provides "recent" semantics that can be reset atomically and efficiently after each `fetch_and_reset()` call.

**Review by acelyc111 — APPROVED** [review]

File: include/dsn/utility/metrics.h [hunk 2]
```diff
@@ -76,20 +76,30 @@
 //   higher performance while consuming more memory if it's updated very frequently.
 // See also include/dsn/utility/long_adder.h for details.
 #define METRIC_DEFINE_counter(entity_type, name, unit, desc, ...)                                  \
-    ::dsn::counter_prototype<::dsn::striped_long_adder> METRIC_##name(                             \
+    ::dsn::counter_prototype<::dsn::striped_long_adder, false> METRIC_##name(                      \
         {#entity_type, #name, unit, desc, ##__VA_ARGS__})
 #define METRIC_DEFINE_concurrent_counter(entity_type, name, unit, desc, ...)                       \
-    ::dsn::counter_prototype<::dsn::concurrent_long_adder> METRIC_##name(                          \
+    ::dsn::counter_prototype<::dsn::concurrent_long_adder, false> METRIC_##name(                   \
```
Comment (levy5307) [inline include/dsn/utility/metrics.h:82]: ::dsn::counter_prototype ==> dsn::counter_prototype

File: include/dsn/utility/metrics.h [hunk 3]
```diff
@@ -343,20 +375,30 @@ class counter : public metric
 
     virtual ~counter() = default;
 
+    long_adder_wrapper<Adder> _adder;
```
Comment (levy5307) [inline include/dsn/utility/metrics.h:378]: why did you make `_addr` public?

Comment (empiredan) [reply include/dsn/utility/metrics.h:378]: > why did you make `_addr` public?

Actually `_addr` is protected, see L373~

**Review by levy5307 — APPROVED** [review]

**Pull request merged**